File: novel-app/server/src/ai/openai.ts
// OpenAI 兼容 provider，照搬 src-tauri/src/ai.rs:259-443。
// 关键契约：
// - 端点 {api_base_url}/v1/chat/completions，Authorization: Bearer {api_key}
// - 请求体仅含单条 user message
// - 非流式：timeout = config.timeout；流式：timeout = config.timeout * 2
// - 重试 3 次，指数退避 2^attempt 秒
// - send 失败总重试；HTTP 非 2xx 仅 5xx 重试，否则 AiFailed("API 返回错误 {status}: {body}")
// - 流式中断：推送 [连接中断/服务端错误，正在重试 (N/M)...] chunk，重新发送整个请求，full_text 保留

import type { AppConfig } from "../config.js";
import { AppError } from "../error.js";
import {
  buildChatRequest,
  backoffDelay,
  MAX_RETRIES,
  type OnChunk,
} from "./index.js";

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

function endpoint(config: AppConfig): string {
  return `${config.api_base_url}/v1/chat/completions`;
}

// 非流式。照搬 ai.rs:259-315
export async function generateOpenai(config: AppConfig, prompt: string): Promise<string> {
  const url = endpoint(config);
  const timeoutMs = config.timeout * 1000;

  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    const body = buildChatRequest(config, prompt, false);
    let resp: Response;
    try {
      resp = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${config.api_key}`,
        },
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(timeoutMs),
      });
    } catch (e) {
      // send 失败
      if (attempt < MAX_RETRIES - 1) {
        await sleep(backoffDelay(attempt));
        continue;
      }
      throw AppError.aiFailed(String(e));
    }

    const status = resp.status;
    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      // 仅 5xx 重试
      if (attempt < MAX_RETRIES - 1 && status >= 500 && status < 600) {
        await sleep(backoffDelay(attempt));
        continue;
      }
      throw AppError.aiFailed(`API 返回错误 ${status}: ${text}`);
    }

    const data = (await resp.json()) as {
      choices?: { message: { content: string } }[];
    };
    const content = data.choices?.[0]?.message?.content;
    if (content === undefined) {
      throw AppError.aiFailed("API 返回空响应");
    }
    return content;
  }
  // 不可达：循环内每个分支都 return/continue/throw
  throw AppError.aiFailed("重试耗尽");
}

// 流式。照搬 ai.rs:317-394
export async function generateStreamingOpenai(
  config: AppConfig,
  prompt: string,
  onChunk: OnChunk,
): Promise<string> {
  const url = endpoint(config);
  const timeoutMs = config.timeout * 2 * 1000;
  let fullText = "";

  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    const body = buildChatRequest(config, prompt, true);
    let resp: Response;
    try {
      resp = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${config.api_key}`,
        },
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(timeoutMs),
      });
    } catch (e) {
      if (attempt < MAX_RETRIES - 1) {
        await sleep(backoffDelay(attempt));
        continue;
      }
      throw AppError.aiFailed(String(e));
    }

    const status = resp.status;
    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      if (attempt < MAX_RETRIES - 1 && status >= 500 && status < 600) {
        onChunk(`\n\n[服务端错误，正在重试 (${attempt + 2}/${MAX_RETRIES})...]\n\n`);
        await sleep(backoffDelay(attempt));
        continue;
      }
      throw AppError.aiFailed(`API 返回错误 ${status}: ${text}`);
    }

    // 流式解析，中途出错则重试
    try {
      await streamOpenaiResponse(resp, (chunk) => {
        fullText += chunk;
        onChunk(chunk);
      });
      return fullText;
    } catch (e) {
      if (attempt < MAX_RETRIES - 1) {
        onChunk(`\n\n[连接中断，正在重试 (${attempt + 2}/${MAX_RETRIES})...]\n\n`);
        await sleep(backoffDelay(attempt));
        continue;
      }
      throw e instanceof Error ? AppError.aiFailed(e.message) : AppError.aiFailed(String(e));
    }
  }
  throw AppError.aiFailed("重试耗尽");
}

// 解析 SSE 流，照搬 ai.rs:396-443
// 按行处理：空行跳过；"data: [DONE]" 结束；"data: {json}" 提取 choices[0].delta.content
async function streamOpenaiResponse(resp: Response, onChunk: OnChunk): Promise<void> {
  if (!resp.body) return;
  const reader = resp.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });

    while (true) {
      const pos = buffer.indexOf("\n");
      if (pos < 0) break;
      const line = buffer.slice(0, pos).trim();
      buffer = buffer.slice(pos + 1);
      if (line === "") continue;
      if (line === "data: [DONE]") return;
      if (!line.startsWith("data: ")) continue;
      const data = line.slice(6);
      let parsed: { choices?: { delta?: { content?: string } }[] };
      try {
        parsed = JSON.parse(data);
      } catch {
        continue;
      }
      const deltaContent = parsed.choices?.[0]?.delta?.content;
      if (deltaContent && deltaContent !== "") {
        onChunk(deltaContent);
      }
    }
  }
}

File: app/src-tauri/src/services/ai/openai.rs
use serde::{Deserialize, Serialize};

use crate::domain::config::AppConfig;
use crate::error::{AppError, Result};
use super::{build_client, Message, retry_delay, retries_exhausted};

const MAX_RETRIES: u32 = 3;

#[derive(Serialize)]
struct OpenAIRequest {
    model: String,
    messages: Vec<Message>,
    stream: bool,
}

#[derive(Deserialize)]
struct OpenAIResponse {
    choices: Vec<OpenAIChoice>,
}

#[derive(Deserialize)]
struct OpenAIChoiceMessage {
    content: String,
}

#[derive(Deserialize)]
struct OpenAIChoice {
    message: OpenAIChoiceMessage,
}

#[derive(Deserialize)]
struct OpenAIDeltaContent {
    content: Option<String>,
}

#[derive(Deserialize)]
struct OpenAIStreamChoice {
    delta: OpenAIDeltaContent,
}

#[derive(Deserialize)]
struct OpenAIStreamChunk {
    choices: Vec<OpenAIStreamChoice>,
}

pub async fn generate(config: &AppConfig, prompt: &str) -> Result<String> {
    let client = build_client(config.timeout)?;
    let url = format!("{}/v1/chat/completions", config.api_base_url);

    for attempt in 0..MAX_RETRIES {
        let request = OpenAIRequest {
            model: config.active_model().to_string(),
            messages: vec![Message {
                role: "user".to_string(),
                content: prompt.to_string(),
            }],
            stream: false,
        };

        let resp = match client
            .post(&url)
            .header("Authorization", format!("Bearer {}", config.api_key))
            .json(&request)
            .send()
            .await
        {
            Ok(r) => r,
            Err(_) if attempt < MAX_RETRIES - 1 => {
                tokio::time::sleep(retry_delay(attempt)).await;
                continue;
            }
            Err(e) => return Err(AppError::AiFailed(e.to_string())),
        };

        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            if attempt < MAX_RETRIES - 1 && status.is_server_error() {
                tokio::time::sleep(retry_delay(attempt)).await;
                continue;
            }
            return Err(AppError::AiFailed(format!(
                "API 返回错误 {}: {}",
                status, body
            )));
        }

        let body: OpenAIResponse = resp.json().await?;
        return body
            .choices
            .into_iter()
            .next()
            .map(|c| c.message.content)
            .ok_or_else(|| AppError::AiFailed("API 返回空响应".to_string()));
    }
    Err(retries_exhausted())
}

pub async fn generate_streaming<F>(config: &AppConfig, prompt: &str, on_chunk: F) -> Result<String>
where
    F: Fn(&str) -> Result<()>,
{
    let client = build_client(config.timeout * 2)?;
    let url = format!("{}/v1/chat/completions", config.api_base_url);

    for attempt in 0..MAX_RETRIES {
        let request = OpenAIRequest {
            model: config.active_model().to_string(),
            messages: vec![Message {
                role: "user".to_string(),
                content: prompt.to_string(),
            }],
            stream: true,
        };

        let resp = match client
            .post(&url)
            .header("Authorization", format!("Bearer {}", config.api_key))
            .json(&request)
            .send()
            .await
        {
            Ok(r) => r,
            Err(_) if attempt < MAX_RETRIES - 1 => {
                tokio::time::sleep(retry_delay(attempt)).await;
                continue;
            }
            Err(e) => return Err(AppError::AiFailed(e.to_string())),
        };

        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            if attempt < MAX_RETRIES - 1 && status.is_server_error() {
                on_chunk(&format!(
                    "\n\n[服务端错误，正在重试 ({}/{})...]\n\n",
                    attempt + 2,
                    MAX_RETRIES
                ))?;
                tokio::time::sleep(retry_delay(attempt)).await;
                continue;
            }
            return Err(AppError::AiFailed(format!(
                "API 返回错误 {}: {}",
                status, body
            )));
        }

        let mut full_text = String::new();
        let mut buffer = String::new();

        match stream_response(resp, &mut buffer, &mut full_text, &on_chunk).await {
            Ok(()) => return Ok(full_text),
            Err(_) if attempt < MAX_RETRIES - 1 => {
                on_chunk(&format!(
                    "\n\n[连接中断，正在重试 ({}/{})...]\n\n",
                    attempt + 2,
                    MAX_RETRIES
                ))?;
                tokio::time::sleep(retry_delay(attempt)).await;
                continue;
            }
            Err(e) => return Err(e),
        }
    }
    Err(retries_exhausted())
}

async fn stream_response<F>(
    resp: reqwest::Response,
    buffer: &mut String,
    full_text: &mut String,
    on_chunk: F,
) -> Result<()>
where
    F: Fn(&str) -> Result<()>,
{
    let mut resp = resp;
    while let Some(chunk) = resp.chunk().await? {
        buffer.push_str(&String::from_utf8_lossy(&chunk));

        while let Some(pos) = buffer.find('\n') {
            let line = buffer[..pos].trim().to_string();
            *buffer = buffer[pos + 1..].to_string();

            if line.is_empty() {
                continue;
            }

            if line == "data: [DONE]" {
                return Ok(());
            }

            let data = match line.strip_prefix("data: ") {
                Some(d) => d,
                None => continue,
            };

            let parsed: OpenAIStreamChunk = match serde_json::from_str(data) {
                Ok(v) => v,
                Err(_) => continue,
            };

            if let Some(choice) = parsed.choices.into_iter().next() {
                if let Some(content) = choice.delta.content {
                    if !content.is_empty() {
                        on_chunk(&content)?;
                        full_text.push_str(&content);
                    }
                }
            }
        }
    }

    Ok(())
}
